app/packages/playback/src/lib/state.ts

```diff
@@ -0,0 +1,384 @@
+import { atom } from "jotai";
+import { atomFamily } from "jotai/utils";
+import { LRUCache } from "lru-cache";
+import { BufferManager, BufferRange } from "../../../utilities/src";
+import {
+  ATOM_FAMILY_CONFIGS_LRU_CACHE_SIZE,
+  DEFAULT_FRAME_NUMBER,
+  DEFAULT_LOOP,
+  DEFAULT_SPEED,
+  DEFAULT_TARGET_FRAME_RATE,
+  DEFAULT_USE_TIME_INDICATOR,
+  LOAD_RANGE_SIZE,
+} from "./constants";
+
+export type PlayheadState =
+  | "playing"
+  | "paused"
+  | "waitingToPlay"
+  | "waitingToPause";
+
+export type TimelineName = string;
+export type FrameNumber = number;
+export type TargetFrameRate = number;
+export type Speed = number;
+export type TotalFrames = number;
+export type TimelineSubscribersMap = Map<
+  SubscriptionId,
+  SequenceTimelineSubscription
+>;
+
+// tood: think about making it a symbol and subscribers a WeakMap
+export type SubscriptionId = string;
+
+export interface SequenceTimelineSubscription {
+  /**
+   * Unique identifier for the subscription.
+   */
+  id: SubscriptionId;
+
+  /**
+   * Fetch and prepare a range of frames.
+   *
+   * Notes:
+   * 1. Subscribers should optimistically load their data as much as possible.
+   * 2. Subscribers should not block rendering while loading data and display a loading indicator.
+   * 3. Subscribers should maintain a buffer of loaded data.
+   * 4. Subscribers should not load data that is already in the buffer.
+   * 5. This function should be referentially stable.
+   *
+   * @param range The range of frames to load.
+   */
+  loadRange: (range: BufferRange) => Promise<void>;
```

Comment:
Awesome! I know we discussed this some offline. Putting some thoughts in writing for visibility...thinking about state with respect to sample data, how much do we want to build into the API?

The current obvious ones are tagging, coloring, sidebar checkboxes.
* Tagging potentially invalidates the `tags` field of frames, etc.
* Checkboxes potentially change overlay/renderable data
* Coloring potentially requires intensive re-processing data (segmentations, heatmaps)

But more generally thinking what limitations we might run into with a single load stage, and how data invalidation works. Or perhaps caching/state changes are up to the implementor of `loadRange`?